python-packaging/execute-code.md

```diff
@@ -0,0 +1,167 @@
+which, in addition to allowing specifying the version of Python, can also read shebang lines and emulate some of that behavior.
+This allows for shebang lines to be re-used between Linux, macOS, and Windows systems. However, on Windows the command must still
+be prefaced with another command (`py`).
+
+```bash
+py my_program.py
+# ✨ Hello from Python ✨
+```
+
+:::{tip}
+While there is no in-source format that can tell Windows what to do with a Python code file, executing a
+Python file with a shebang on Windows also does not cause any issues. Python just sees the whole line as
+a comment and ignores it! So even if you develop on Windows, it may be a good idea to add the shebang as
+the first line of your scripts, so that colleagues on different systems can also run it.
+:::
+
+## Executable comparisons
+
+### Pros of passing a file to `python`:
+- don't need execute permissions
+- works for every system
+- explicit about what you expect to happen
+
+### Pros of inserting a shebang to the file:
+- file is associated with specific python
+  - don't have to remember which
+- don't have to use the `python` command
+  - don't have to even remember it is a Python script
+
+
+# Share Code
+
+## Execute a python package
+
+In [Code Workflow Logic][Code Workflow Logic] you learned of the two primary ways to execute a stand-alone Python script.
+There are two other ways to execute Python as commands, both of which work for code that has been formatted as a package.
+
+### Entrypoints
+
+There is a special `entrypoint` a package can specify in its configuration which will direct installers to create an
+executable command. Entrypoints are a general purpose plug-in system for Python packages, but the
+[`console_scripts`](https://packaging.python.org/en/latest/specifications/entry-points/#use-for-scripts)
+entry is specifically targeted at creating executable commands on systems that install the package.
+
+The target of a `scripts` definition should be one function within your package, which will be directly executed
+when the command is invoked in the shell. A `scripts` definition in your `pyproject.toml` looks like:
+
+```toml
+[project.scripts]
+COMMAND = "my_package.my_module:my_function"
+```
+
+where `COMMAND` is the name of the command that will be made available after installation, `my_package` is the name of
+your top-level package import, `my_module` is the name of any sub-modules in your package (optional, or may be
+repeated as necessary to access the correct sub-module), and `my_funciton` is the function that will be called
+(without parameters) when the command is invoked.
+
+Scripts defined in project configuration, such as `pyproject.toml`, do not need to exist as independent files in
+the package repository, but will be created by installation tools, such as `pip`, at the time the package is
+installed, in a manner customized to the current operating system.
```

Comment:
i am not sure what this means?

Reply:
scripts are not files in the repo's source. The `black` "command" is constructed and saved during install, it is part of the job of an installer to make these scripts. This is as opposed to keeping explicit scripts checked into the VCS (like a `bin/run_me` with a shebang) that some projects do. Some even include those scripts in the sdist/wheel; if they do, they also get installed into e.g. the venv and appear on PATH. But those are not "entrypoint scripts".

Reply:
aha that's what i figured you were referring to - i think this might be TMI for newbies :). it makes me think like "wait so should I not put this `.py` file in my repo? how will the tool know how to generate it for me if it's not there? when what you're referring to are the little autogenerated shim files that just call the script with shebang/etc. 

Might be another good one for an admonition box after explaining the concept of `[project.scripts]` kind of scripts - "try this: install your package in a virtual environment, and then open the `venv/bin/your_script` file - what the heck is that about!?!?!" 

i think it's important information (in the same way that i always tell anyone who wants to know what a virtual environment is to just read the `bin/activate` file (actually usually i parse it for them because there's really only two things you need to know but they're hidden in the shell script)) but  might be a bit much in main text, and seems like it fits as a "if you want to know more" for the curious campers <3

Reply:
Took this out for now